web3/conversions.nim

```diff
@@ -31,6 +31,12 @@ template invalidQuantityPrefix(s: string): bool =
 
 func `%`*(n: Int256|UInt256): JsonNode = %("0x" & n.toHex)
 
+func safelyParseHexU256(hexStr: string, argName: string): UInt256 =
+  try:
+    return hexStr.parse(StUint[256], 16)
+  except AssertionError:
```

Comment:
You should not catch `Defects` like this (`AssertionError` is a `Defect`). If the `parse` string has a pre-condition that the input must be a valid hex string, you should probably raise an issue in the upstream library as such a requirement is not quite reasonable for a parsing function.

In the meantime, you can consider using a different hex-parsing function or validating the input with a helper function before calling `parse`. 

Defects should not be handled in general, because by definition they are unexpected events that should never happen in the code. The distinction between Defects and CatchableErrors is explained here:

https://gist.github.com/zah/d2d729b39d95a1dfedf8183ca35043b3

Reply:
Hmm, I move this to a function because of the `TODO` which was present in this file before:

https://github.com/status-im/nim-web3/blame/577c86e34bd613278912d72d43e3bf398e1e1d7c/web3/conversions.nim#L43-44

Reply:
Actually there was no crash in this area, but I thought it can be hardened also, I should revert this and re-apply the `TODO`?

Reply:
As seen from the lines below, a simple JSON object can indeed cause the parser to crash if we do not handle `AssertionError` in the event of a malformed input string:

```nim
    should_be_value_error("0x", resRlpEncodedBytes)
    should_be_value_error("0x", resTypedTransaction)
```

What do you recommend we do about such cases?

Reply:
Defects are run-time problems arising from code that violates some documented usage requirements of a particular library (I referred to these as "pre-conditions" in the previous message). In other words, these are errors for which the correct solution is to change the code in a way that satisfies the requirements of the library and makes the defect impossible to arise. In contrast, `CatchableErrors` are expected to arise even in a correctly implemented program because they are a natural consequence of the run-time environment of the program (e.g. a network connection may go down, the user may enter an invalid value for some of the program inputs, etc).

As a concrete example, a defect would be accessing an element of a sequence that's outside of its bounds. The solution would be to change the code, such that you are checking that the index is a natural number below the `len` of the sequence before you access the element.

The `hexToByteArray` API should never fail with an `AssertionError` in the way you use it below. Some of its alternative APIs that accept `from` and `to` parameters can fail with an `AssertionError` if these parameters are given illegal values.

The run-time errors that must be handled here are invalid hex strings (i.e. ones containing non-hex characters) and these will produce a `ValueError` in the library (which is not a defect, but rather a `CatchableError`, so it must be handled by a correctly written program at run-time).

Reply:
Yep I see that makes sense about the overflow situation, however the `hexToByteArray` is actually throwing an `AssertionError` when using `0x` as an input string.

Do you mean I should write a step-by-step validation function to ensure we're actually using hexadecimal strings?

It seems that catching the `AssertionError` and pushing value error would be much more optimised, or am I missing something?

Reply:
I moved this comment to the relevant (not outdated section)